### ISSUE TITLE
open txt file and play wav file

### ISSUE THREAD

**Issue by genia1991** [TMRh20/TMRpcm#133]
Hi everyone. I have an issue trying to open not only wav file. I wanna open txt file parallelly. Is there any how to do that ?
useing just SD library. Thanks for help
here is my sketch:

`#include <SD.h>                      // need to include the SD library
#define SD_ChipSelectPin 10  //using digital pin 4 on arduino nano 328, can use other pins
#include <TMRpcm.h>           //  also need to include this library...
#include <SPI.h>

TMRpcm tmrpcm;   // create an object for use in this sketch

File myFile;
void setup(){
  tmrpcm.speakerPin = 9; //5,6,11 or 46 on Mega, 9 on Uno, Nano, etc

  Serial.begin(9600);
  if (!SD.begin(SD_ChipSelectPin)) {  // see if the card is present and can be initialized:
    Serial.println("SD fail");  
    return;   // don't do anything more if not
  }
  tmrpcm.play("22.wav"); //the sound file "music" will play each time the arduino powers up, or is reset

}

void loop(){  

  if(Serial.available()){    
    if(Serial.read() == 'd'){ //send the letter p over the serial monitor to start playback
      tmrpcm.play("22.wav");
    }
  }
  //--------------------------------------------------------
  // re-open the file for reading:
  myFile = SD.open("test.txt");
  if (myFile) {
    Serial.println("test.txt:");

    // read from the file until there's nothing else in it:
    while (myFile.available()) {
      Serial.write(myFile.read());
    }
    // close the file:
    myFile.close();
  } else {
    // if the file didn't open, print an error:
    Serial.println("error opening test.txt");
  }
  
//--------------------------------------------------------
}`

**Comment by TMRh20** [comment]
Can't currently access the SD card while a file is playing, but if you use `tmrpcm.isPlaying();` to prevent opening while playing a file, or `tmrpcm.stopPlayback();` to stop playback before accessing SD, then it would work.

I'm not sure why this doesn't work, since the library can open and read multiple files for multi-track playback, but opening one from the sketch causes issues. I'll have to look into it further.

**Comment by TMRh20** [comment]
In looking at this further I am fairly certain it has to do with the core functionality of the library.

AVR micro-controllers are typically single threaded, so even though a user may open 2 files from SD card, only one of them will be accessed at any given time.

TMRpcm uses interrupts to achieve asynchronous playback, and even when two files are open for playback, they are never accessed at the same time. 

When opening a file from the main sketch, if audio is playing, it will literally interrupt the code execution of the main loop, trying to read from file at the same time another file is open/being accessed. The SD lib is not designed for this and it won't work.

tl;dr: In short, this will never work during playback.

**Comment by TMRh20** [comment]
Just FYI I was reminded of my own library Auto Analog Audio, which would allow this because it can play audio without using interrupts. 

http://tmrh20.github.io/AAAudio/SDAudioBasic_8ino-example.html

**Comment by genia1991** [comment]
I am very appreciate for your help and I find every your reply very useful. This helped me to solve the problem :)
